### PR TITLE
adds dns support for ns1 provider

### PR DESCRIPTION
**Pull request by piratimir**
The methods that deal with records are not finished yet.

### CODE REVIEW

**Comment by tonybaloney** [comment]
Thanks @jetbird for the PR. Let me know when this is completed. Also if you install flake8 locally you can run the code checks to pass the failing lint build job in travis.

**Comment by piratimir** [comment]
@tonybaloney @Kami   I think we have a problem in implementing the def create_record(self, name, zone, type, data, extra=None) method. From what I understand when reading the API, the provider does not support adding a name for the new record.

Under Records take a look on the PUT method https://ns1.com/api/

**Comment by Kami** [comment]
@jetbird I checked the docs and it looks like you need to provide a fully qualified record name when creating a record.

This means record name + domain name. It looks like that in the docs they refer to the record name + domainname  as "domain".

For example if record name is `foo` and domain name is `example.com` you would need to pass `example.com` for the zone and `foo.example.com` for domain.

**Comment by piratimir** [comment]
@Kami, you are right. Coding it now.

**Comment by piratimir** [comment]
@tonybaloney  It is finished man. Please take a look at the code.

**Comment by tonybaloney** [comment]
@jetbird the lint is failing:
libcloud/common/nsone.py:20:80: E501 line too long (85 > 79 characters)
libcloud/common/nsone.py:23:80: E501 line too long (88 > 79 characters)
libcloud/dns/drivers/nsone.py:7:80: E501 line too long (80 > 79 characters)
libcloud/dns/drivers/nsone.py:8:80: E501 line too long (114 > 79 characters)
libcloud/dns/drivers/nsone.py:60:80: E501 line too long (90 > 79 characters)
libcloud/dns/drivers/nsone.py:79:80: E501 line too long (115 > 79 characters)
libcloud/dns/drivers/nsone.py:90:80: E501 line too long (91 > 79 characters)
libcloud/dns/drivers/nsone.py:93:80: E501 line too long (90 > 79 characters)
libcloud/dns/drivers/nsone.py:129:80: E501 line too long (90 > 79 characters)
libcloud/dns/drivers/nsone.py:132:80: E501 line too long (85 > 79 characters)
libcloud/dns/drivers/nsone.py:138:80: E501 line too long (92 > 79 characters)
libcloud/dns/drivers/nsone.py:140:80: E501 line too long (80 > 79 characters)
libcloud/dns/drivers/nsone.py:149:80: E501 line too long (96 > 79 characters)
libcloud/dns/drivers/nsone.py:164:80: E501 line too long (86 > 79 characters)
libcloud/dns/drivers/nsone.py:169:80: E501 line too long (96 > 79 characters)
libcloud/dns/drivers/nsone.py:189:80: E501 line too long (90 > 79 characters)
libcloud/dns/drivers/nsone.py:205:80: E501 line too long (91 > 79 characters)
libcloud/dns/drivers/nsone.py:208:80: E501 line too long (90 > 79 characters)
libcloud/dns/drivers/nsone.py:230:80: E501 line too long (90 > 79 characters)
libcloud/dns/drivers/nsone.py:243:80: E501 line too long (92 > 79 characters)
libcloud/dns/drivers/nsone.py:246:80: E501 line too long (89 > 79 characters)
libcloud/dns/drivers/nsone.py:277:80: E501 line too long (113 > 79 characters)

**Comment by piratimir** [comment]
@tonybaloney I managed to flake8 the code.

**Comment by Kami** [comment]
@jetbird Nice work so far :+1: 

The next step would be adding some tests and documentation :)

File: libcloud/dns/drivers/nsone.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+        for zone in zones_list:
+            zone_ids.append(zone.id)
+
+        return zone_id in zone_ids
+
+    def _to_zone(self, item):
+        common_attr = ['zone', 'id', 'type']
+        extra = {}
+        for key in item:
+            if key not in common_attr:
```
Comment (Kami) [inline libcloud/dns/drivers/nsone.py:308]: I believe this should be `if key in common_attr` (dropping `not`)?

File: libcloud/dns/drivers/nsone.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+        zone_ids = []
+        for zone in zones_list:
+            zone_ids.append(zone.id)
+
+        return zone_id in zone_ids
+
+    def _to_zone(self, item):
+        common_attr = ['zone', 'id', 'type']
+        extra = {}
+        for key in item:
```
Comment (Kami) [inline libcloud/dns/drivers/nsone.py]: I also assume this should be `for key in item.keys():`?

File: libcloud/dns/drivers/nsone.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+        zones = []
+        for item in items:
+            zones.append(self._to_zone(item))
+
+        return zones
+
+    def _to_record(self, item, zone):
+        common_attr = ['id', 'short_answers', 'answers', 'domain', 'type']
+        extra = {}
+        for key in item:
```
Comment (Kami) [inline libcloud/dns/drivers/nsone.py]: Same here.

File: libcloud/dns/drivers/nsone.py (hunk 1 above)
Comment (piratimir) [inline libcloud/dns/drivers/nsone.py:308]: If a key is not in common_attr list then we add it in the extra dict as a key with the item.get('key') as value. The keys in common_attr are needed to build the Zone object. Why do you think we should drop it?

File: libcloud/dns/drivers/nsone.py (hunk 3 above)
Comment (piratimir) [inline libcloud/dns/drivers/nsone.py]: Done :) :+1:

**Comment by piratimir** [comment]
@Kami @tonybaloney  guys did you find some time to check the tests?

File: libcloud/test/dns/test_nsone.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
+import sys
+import unittest
+
+from libcloud.test import MockHttp
+from libcloud.test.file_fixtures import DNSFileFixtures
+from libcloud.test.secrets import DNS_PARAMS_NSONE
```
Comment (tonybaloney) [inline libcloud/test/dns/test_nsone.py:6]: This does not exist, you need to update test/secrets-dist.py

Comment (piratimir) [inline libcloud/test/dns/test_nsone.py:6]: @tonybaloney I thought DNS_PARAMS_NSONE is supposed to stay locally?

Comment (piratimir) [inline libcloud/test/dns/test_nsone.py:6]: @tonybaloney I  updated and pushed secrets.py. I do not know why the travis builds are failing. The tests pass when run on my local machine.

**Comment by tonybaloney** [comment]
@jetbird secrets.py is on the ignore list, part of the build is to copy secrets-dist.py to secrets.py. You need to apply your changes to secrets-dist and apply those changes. Not secrets.py

**Comment by piratimir** [comment]
@tonybaloney Done.

**Comment by tonybaloney** [comment]
Excellent, thats the one. If CI gives the :green_apple: then I will merge. LGTM

**Comment by piratimir** [comment]
@tonybaloney It does not look like that. I  also updated the fixtures with blank lines at each one of them, but when pushing on github they get ignored.

**Comment by tonybaloney** [comment]
@jetbird `except NsOneException, e:` is not a valid syntax in Python 3. 

Do you have Tox? This would really help you testing before you commit.

**Comment by piratimir** [comment]
@tonybaloney I will install it now.

**Comment by piratimir** [comment]
@tonybaloney I installed tox and ran the tests with sudo tox, but the problem is that many interpreters are not found. I am not familiar with tox, so I guess I need to read the docs. Where should I install these interpreters?

Is the following the only problem or are there more. Because if it is the only one I can fix this and take care later about the tox?

except NsOneException, e: is not a valid syntax in Python 3.

**Comment by tonybaloney** [comment]
@jetbird you can review the errors here  https://travis-ci.org/apache/libcloud/builds/116256818

**Comment by piratimir** [comment]
@tonybaloney I decided to follow this approach to keep one code base through different versions of python.

```
  try:
       response = self.connection.request(action=action, method='GET')
  except NsOneException:
       e = sys.exc_info()[1]
```

Now will try to pass the tests with tox.

**Comment by tonybaloney** [comment]
Use `tox -e py34` as an example to just test 3.4. You don't need to install every possible version, just at least 2.7 and 3.4+.

**Comment by jetbird** [comment]
@tonybaloney  Thanks for being so patient and helping :) But where should these versions of python be installed? Do I need to create new virtual environments and install them in there?

**Comment by jetbird** [comment]
I just installed python3.4 on my machine.

Here is how my tox.ini file looks.
`[tox]
envlist = py{2.5,2.6,2.7,pypy,pypy3,3.2,3.3,3.4,3.5},lint

[testenv]
deps =
    -r{toxinidir}/requirements-tests.txt
    lockfile
    py{2.5,2.6,2.7}: paramiko
    py{2.5,2.6}: unittest2
commands = cp libcloud/test/secrets.py-dist libcloud/test/secrets.py
           python setup.py test
basepython =
    py2.5: python2.5
    py2.6: python2.6
    {py2.7,lint,pylint,docs}: python2.7
    pypypy: pypy
    pypypy3: pypy3
    py3.2: python3.2
    py3.3: python3.3
    py3.4: python3.4
    py3.5: python3.5
whitelist_externals = cp

[testenv:docs]
deps = sphinx
       pysphere
       backports.ssl_match_hostname
changedir = docs
commands = python ../contrib/generate_provider_feature_matrix_table.py
           sphinx-apidoc -d 4 ../libcloud/ -o apidocs/
           sphinx-build -W -b html -d {envtmpdir}/doctrees . _build/html

[testenv:docs-travis]
# Note: We don't build API docs on Travis since it causes build failures because
# those API docs files are not included anywhere.
deps = sphinx
       pysphere
       backports.ssl_match_hostname
changedir = docs
commands = python ../contrib/generate_provider_feature_matrix_table.py
           sphinx-build -W -b html -d {envtmpdir}/doctrees . _build/html

[testenv:scrape-ec2-prices]
deps = requests
       demjson
commands = python contrib/scrape-ec2-prices.py

[testenv:pylint]
deps = pylint
       backports.ssl_match_hostname
commands = pylint --rcfile=.pylintrc -E libcloud/

[testenv:lint]
deps = -r{toxinidir}/requirements-tests.txt
       backports.ssl_match_hostname

commands = flake8 --ignore=E402 --exclude="test" libcloud/
           flake8 --ignore=E402 --max-line-length=160 libcloud/test/
           flake8 --ignore=E402 demos/
           flake8 --ignore=E402,E902 docs/examples/
           flake8 --ignore=E402,E902 --max-line-length=160 contrib/
           python -mjson.tool libcloud/data/pricing.json
`

**Comment by piratimir** [comment]
@tonybaloney I get the following output when running the command.

```
lostbird@lostbird:~/projects/libcloud$ sudo tox -e py34
GLOB sdist-make: /home/lostbird/projects/libcloud/setup.py
py34 create: /home/lostbird/projects/libcloud/.tox/py34
ERROR: InterpreterNotFound: 
____________________________________________________________________________ summary _____________________________________________________________________________
ERROR:   py34: InterpreterNotFound: 

```

**Comment by piratimir** [comment]
@tonybaloney  I managed to run the tests against python3.4 and they pass. The only check that is failing on travis is the lint check. Maybe it is the fixtures, I have no idea. Thanks for being so patient and helping.

https://travis-ci.org/apache/libcloud/builds/116270772

**Comment by piratimir** [comment]
@Kami @tonybaloney Hey guys when using flake8 on nsone module and test_nsone I do not get any errors. Why do you think the lint build job in travis is failing?

**Comment by piratimir** [comment]
@tonybaloney @Kami any update on this please?

File: libcloud/test/dns/test_nsone.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+from libcloud.dns.drivers.nsone import NsOneDNSDriver
+from libcloud.utils.py3 import httplib
+from libcloud.dns.types import ZoneDoesNotExistError, ZoneAlreadyExistsError,\
+    RecordDoesNotExistError,  RecordType
```
Comment (tonybaloney) [inline libcloud/test/dns/test_nsone.py]: There are 2 spaces after the comma. this is failing the lint. remove one of the spaces and just commit that change then we're good to go @jetbird

Comment (piratimir) [inline libcloud/test/dns/test_nsone.py]: Ok man :) :+1:

**Comment by piratimir** [comment]
@tonybaloney I removed the extra spaces, but it still fails.

**Comment by tonybaloney** [comment]
@jetbird so close!! libcloud/test/dns/test_nsone.py:10:2: E121 continuation line under-indented for hanging indent

**Comment by tonybaloney** [comment]
yeah!!

**Comment by piratimir** [comment]
@tonybaloney finally :dancers:

**Comment by tonybaloney** [comment]
Merged. Thanks @jetbird it might be time to get a Python editor with a built in linter! Try Komodo or Atom :-)

**Comment by piratimir** [comment]
@tonybaloney Thanks for the help and support :+1: I am using pycharm and flake8 is not working ok. I installed pylint via pip. I will be contributing again, so please if you can give me your email address would be great :)

**Pull request closed without merging**